### PR TITLE
2.1.0 social

### PR DESCRIPTION
**Pull request by ficiverson**
Social integration with capabilities for login/sign in with HALO, Facebook or Google plus.

### CODE REVIEW

File: sdk-libs/halo-social/build.gradle [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+dependencies {
+    compile "com.mobgen.halo.android:halo-sdk:${rootProject.ext.haloVersion}"
+    provided "com.google.android.gms:play-services-auth:${rootProject.ext.firebaseVersion}"
+    provided "com.facebook.android:facebook-android-sdk:4.13.1"
+    apt "com.bluelinelabs:logansquare-compiler:${rootProject.ext.loganSquareVersion}"
+
+    //Testing dependencies
+    testApt "com.bluelinelabs:logansquare-compiler:${rootProject.ext.loganSquareVersion}"
+    testCompile "com.mobgen.halo.android:halo-testing:${rootProject.ext.haloVersion}"
+    testCompile "com.squareup.okhttp3:mockwebserver:${mockWebServerVersion}"
```
Comment (droidpl) [inline sdk-libs/halo-social/build.gradle]: I think this is a dependency taken from halo-testing

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+     * @param accountType The account type to store on account manager.
+     * @param tokenType   The token type
+     * @param userName    The username of the user.
+     * @param password    The password of the user.
+     * @param authToken The token generated or updated
+     *
+     * @return Return the account added.
+     */
+    @Nullable
+    public Account addAccountWithPassword(@NonNull String accountType, @NonNull String tokenType, @NonNull String userName, @NonNull String password, @NonNull String authToken) {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java]: Add also token type types into this class

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        AssertionUtils.notNull(accountType, "accountType");
+        AssertionUtils.notNull(tokenType, "socialToken");
+        AssertionUtils.notNull(userName, "userName");
+        AssertionUtils.notNull(password, "password");
+        if (ActivityCompat.checkSelfPermission(Halo.instance().context(), Manifest.permission.GET_ACCOUNTS) != PackageManager.PERMISSION_GRANTED) {
+            Account account = new Account(userName, accountType);
+            if(checkAccountExist(account,accountType)){
+                mAccountManager.setPassword(account,password);
+                mAccountManager.setUserData(account,tokenType,authToken);
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java]: Better namings for tokentypes

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+            if(checkAccountExist(account,accountType)){
+                mAccountManager.setUserData(account, tokenType, authToken);
+                //if we have a halo credential on the account we will restore with username/password
+                if(mAccountManager.getPassword(account)==null) {
+                    mAccountManager.setUserData(account, TOKEN_TYPE, tokenType);
+                }
+                mAccountManager.setUserData(account, AuthTokenType.HALO_AUTH_TOKEN, haloAccessToken);
+                return setAuthToken(account,tokenType,authToken);
+            }else {
+                deleteOtherAccounts(account,accountType);
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java]: Does not fit DRY

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+     * Gets the account stored based on account type and token type. There is only one account on account manager
+     *
+     * @param accountType The account type.
+     *
+     * @return Return the account stored.
+     */
+    @Nullable
+    public Account recoverAccount(@NonNull String accountType) {
+        AssertionUtils.notNull(accountType, "accountType");
+        if (ActivityCompat.checkSelfPermission(Halo.instance().context(), Manifest.permission.GET_ACCOUNTS) != PackageManager.PERMISSION_GRANTED) {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java]: Not needed

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+     *
+     * @param account The account.
+     * @param accountType The account type to authenticate.
+     * @return True if account is already authenticated, false otherwise.
+     */
+    @NonNull
+    private boolean checkAccountExist(@NonNull Account account, @NonNull String accountType) {
+        AssertionUtils.notNull(account, "account");
+        AssertionUtils.notNull(accountType, "accountType");
+        if (ActivityCompat.checkSelfPermission(Halo.instance().context(), Manifest.permission.GET_ACCOUNTS) != PackageManager.PERMISSION_GRANTED) {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java]: Not needed

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+     * @param account The account.
+     * @param accountType The account type.
+     *
+     * @return True if account is already authenticated, false otherwise.
+     */
+    @NonNull
+    private void deleteOtherAccounts(@NonNull Account account,@NonNull String accountType) {
+        AssertionUtils.notNull(account, "account");
+        AssertionUtils.notNull(accountType, "account");
+        if (ActivityCompat.checkSelfPermission(Halo.instance().context(), Manifest.permission.GET_ACCOUNTS) != PackageManager.PERMISSION_GRANTED) {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AccountManagerHelper.java]: Not needed

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AuthTokenType.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package com.mobgen.halo.android.social.authenticator;
+
+
+public class AuthTokenType {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/authenticator/AuthTokenType.java]: Merge with helper

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/login/LoginInteractor.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    private String mUserName;
+    /**
+     * User password.
+     */
+    private String mPassword;
+    /**
+     * Device.
+     */
+    private Device mDevice;
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/login/LoginInteractor.java]: Use the alias

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/login/LoginRepository.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Provides the user identity with a social token.
+     *
+     * @param socialApiName The remote data source.
+     * @param socialToken The remote data source.
+     * @param device The remote data source.
+     * @return The user identified
+     */
+    @NonNull
+    public synchronized IdentifiedUser loginSocialProvider(@NonNull String socialApiName, @NonNull String socialToken, @NonNull Device device) throws HaloNetException {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/login/LoginRepository.java]: Remove synchronized

Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/login/LoginRepository.java]: Return values from repositories must be HaloResultV2<Something>

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/models/HaloAuthProfile.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    String mPassword;
+    /**
+     * The device alias.
+     */
+    @Nullable
+    @JsonField(name= "deviceId")
+    String mAlias;
+
+
+    public HaloAuthProfile(){}
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/models/HaloAuthProfile.java]: Add a factory method create

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/models/HaloUserProfile.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+     */
+    @Nullable
+    @JsonIgnore(ignorePolicy = JsonIgnore.IgnorePolicy.SERIALIZE_ONLY)
+    @JsonField(name = "id")
+    String mIdentifiedId;
+    /**
+     * The email.
+     */
+    @Nullable
+    @JsonIgnore(ignorePolicy = JsonIgnore.IgnorePolicy.SERIALIZE_ONLY)
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/models/HaloUserProfile.java]: Lets discuss this internally

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/models/Token.java [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+import java.util.Date;
+
+
+/**
+ * Token used for the OAuth protocol in HALO.
+ */
+@Keep
+@JsonObject
+public class Token implements Parcelable {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/models/Token.java]: Repeated model. It is in the halo-sdk

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/providers/SocialProvider.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    boolean linkedAppAvailable(@NonNull Context context);
+
+    /**
+     * Authenticates using the halo instance and providing the result in the callback.
+     *
+     * @param halo The halo instance.
+     * @param callback The callback.
+     */
+    void authenticate(@NonNull Halo halo, @NonNull CallbackV2<HaloSocialProfile> callback);
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/providers/SocialProvider.java]: Modify this signature to allow the auth profile

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/providers/facebook/FacebookSocialProvider.java [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    public boolean linkedAppAvailable(@NonNull Context context) {
+        return true;
+    }
+
+    @Override
+    public void authenticate(final @NonNull Halo halo, @NonNull CallbackV2<HaloSocialProfile> callback) {
+        final Subscriber subscriber = this;
+        initIfNeeded(halo.context());
+        mSocialApi = (HaloSocialApi)halo.manager().haloSocial();
+        final String authSocialToken =  mSocialApi.recoverAuthToken(AuthTokenType.FACEBOOK_AUTH_TOKEN);
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/providers/facebook/FacebookSocialProvider.java]: Reduce complexity

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/providers/facebook/FacebookSocialProvider.java [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+            haloSocialProfileHaloResult = error(resultData);
+        }
+
+        //Result must not be null, some error in configuration
+        if (haloSocialProfileHaloResult == null) {
+            String msg = "This should never happen. An event must have a FACEBOOK_SIGN_IN_RESULT. Contact the halo support.";
+            Halog.wtf(getClass(), msg);
+            throw new IllegalStateException(msg);
+        }
+        //login user into halo with social credentials or notified process ended if we cant obtain social provider social token
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/providers/facebook/FacebookSocialProvider.java]: Reduce complexity

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/register/RegisterRemoteDatasource.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+     * @param haloAuthProfile The auth profile.
+     * @param haloUserProfile The user profile.
+     * @throws HaloNetException Networking exception.
+     * @return The user identity as IdentifiedUser
+     */
+    @NonNull
+    public HaloUserProfile register(@NonNull HaloAuthProfile haloAuthProfile, @NonNull HaloUserProfile haloUserProfile) throws HaloNetException {
+        JSONObject jsonObject = null;
+        try {
+            Register register = new Register(haloAuthProfile, haloUserProfile);
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/register/RegisterRemoteDatasource.java]: Use the halo.framework().parser()

File: sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/register/RegisterRepository.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Sign in a user in halo
+     *
+     * @param haloAuthProfile The auth profile
+     * @param haloUserProfile The user profile
+     * @return The user identified
+     * @throws HaloNetException Networking exception.
+     */
+    @NonNull
+    public synchronized HaloUserProfile registerHalo(@NonNull HaloAuthProfile haloAuthProfile, @NonNull HaloUserProfile haloUserProfile) throws HaloNetException {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/java/com/mobgen/halo/android/social/register/RegisterRepository.java]: HaloResult

File: sdk-libs/halo-social/src/main/res/xml/halo_authenticator.xml [hunk 19]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<account-authenticator xmlns:android="http://schemas.android.com/apk/res/android"
+    android:accountType="halo.account.manager"
+    android:icon="@drawable/myicon"
+    android:smallIcon="@drawable/myicon"
+    android:label="Halo Social APIß"
```
Comment (droidpl) [inline sdk-libs/halo-social/src/main/res/xml/halo_authenticator.xml]: Remove the beta char

File: sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/HaloSocialApiTest.java [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+
+    public void givenPermissions(){
+        mContext = mock(Context.class);
+        when(mContext.checkPermission(eq("AUTHENTICATE_ACCOUNTS"),anyInt(),anyInt()))
+                .thenReturn(PackageManager.PERMISSION_GRANTED);
+        when(mContext.checkPermission(eq("GET_ACCOUNTS"),anyInt(),anyInt()))
+                .thenReturn(PackageManager.PERMISSION_GRANTED);
+        when(mContext.checkPermission(eq("MANAGE_ACCOUNTS"),anyInt(),anyInt()))
+                .thenReturn(PackageManager.PERMISSION_GRANTED);
+        when(mContext.checkPermission(eq("WRITE_SYNC_SETTINGS"),anyInt(),anyInt()))
```
Comment (droidpl) [inline sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/HaloSocialApiTest.java]: Check this permission

File: sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/models/TokenTest.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+import com.mobgen.halo.android.social.models.Token;
+import com.mobgen.halo.android.testing.HaloRobolectricTest;
+import com.mobgen.halo.android.testing.TestUtils;
+
+import org.junit.Test;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
+
+
+public class TokenTest extends HaloRobolectricTest {
```
Comment (droidpl) [inline sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/models/TokenTest.java]: Not needed

File: sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/providers/SocialNotAvailableExceptionTest.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Test
+    public void thatSocialNeExceptionIsThrow(){
+        mHaloSocialApi.release();
+        if(!mHaloSocialApi.isSocialNetworkAvailable(HaloSocialApi.SOCIAL_FACEBOOK)){
+            try {
+                mHaloSocialApi.login(HaloSocialApi.SOCIAL_FACEBOOK, new CallbackV2<HaloSocialProfile>() {
+                    @Override
+                    public void onFinish(@NonNull HaloResultV2<HaloSocialProfile> result) {
+
```
Comment (droidpl) [inline sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/providers/SocialNotAvailableExceptionTest.java]: Check this is called or send null or ensure not called. But not empty

File: sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/providers/SocialNotAvailableExceptionTest.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Test
+    public void thatCreateANewSocialNotAvailableExceptionWithErrorCausing() {
+        SocialNotAvailableException socialException = new SocialNotAvailableException("Social provider not exist", new Exception());
+        assertThat(socialException).isNotNull();
+        assertThat(socialException.getMessage()).contains("Social");
+    }
+
+    @Test
```
Comment (droidpl) [inline sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/providers/SocialNotAvailableExceptionTest.java]: Expected exception type

File: sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/providers/facebook/FacebookSocialProviderTest.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+
+import static com.mobgen.halo.android.testing.CallbackFlag.newCallbackFlag;
+import static org.assertj.core.api.Java6Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@Config(shadows = StringShadowResources.class)
+public class FacebookSocialProviderTest extends HaloRobolectricTest {
+
+    @Override
```
Comment (droidpl) [inline sdk-libs/halo-social/src/test/java/com/mobgen/halo/android/social/social/providers/facebook/FacebookSocialProviderTest.java]: Not needed method

File: sdk/halo-framework/build.gradle [hunk 25]
```diff
@@ -25,10 +25,6 @@ android {
 
     testOptions {
         unitTests.all {
```
Comment (droidpl) [inline sdk/halo-framework/build.gradle]: Weird remove

File: sdk/halo-framework/src/test/java/com/mobgen/halo/android/framework/mock/NetworkActionMocker.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
+//package com.mobgen.halo.android.framework.mock;
```
Comment (droidpl) [inline sdk/halo-framework/src/test/java/com/mobgen/halo/android/framework/mock/NetworkActionMocker.java]: Check this

File: sdk/halo-framework/src/test/java/com/mobgen/halo/android/framework/mock/StorageActionMocker.java [hunk 27]
```diff
@@ -0,0 +1,39 @@
+//package com.mobgen.halo.android.framework.mock;
```
Comment (droidpl) [inline sdk/halo-framework/src/test/java/com/mobgen/halo/android/framework/mock/StorageActionMocker.java]: Check this

File: sdk/halo-framework/src/test/java/com/mobgen/halo/android/framework/mock/instrumentation/HaloDatabaseInstrument.java [hunk 28]
```diff
@@ -1,6 +1,5 @@
 package com.mobgen.halo.android.framework.mock.instrumentation;
```
Comment (droidpl) [inline sdk/halo-framework/src/test/java/com/mobgen/halo/android/framework/mock/instrumentation/HaloDatabaseInstrument.java]: Check this

File: sdk-libs/halo-social/proguard-rules.pro [hunk 29]
```diff
@@ -0,0 +1,7 @@
+-keepattributes Exceptions
+-keepparameternames
+-keepattributes EnclosingMethod
+
+-keep class com.bluelinelabs.logansquare.** { *; }
+-keep @com.bluelinelabs.logansquare.annotation.JsonObject class *
```
Comment (ficiverson) [inline sdk-libs/halo-social/proguard-rules.pro]: According to official documentation pages there is no need to add proguard rules:
https://developers.facebook.com/docs/android/getting-started
https://developers.google.com/android/guides/setup

**Pull request merged**